File: src/main/java/com/openvalue/boardgameratings/service/rating/BoardGameRatingService.java
package com.openvalue.boardgameratings.service.rating;

import com.openvalue.boardgameratings.api.*;
import com.openvalue.boardgameratings.service.boardgame.BoardGameEntity;
import com.openvalue.boardgameratings.service.boardgame.BoardGameNotFound;
import com.openvalue.boardgameratings.service.boardgame.BoardGameRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@Service
public class BoardGameRatingService {
    private final BoardGameRepository boardGameRepository;
    private final RateRepository rateRepository;

    @Autowired
    public BoardGameRatingService(BoardGameRepository boardGameRepository, RateRepository rateRepository) {
        this.boardGameRepository = boardGameRepository;
        this.rateRepository = rateRepository;
    }

    BoardGame ratingBoardGame(RateBoardGame rateBoardGame) throws BoardGameNotFound {
        final RateEntity build = new RateEntity(null, rateBoardGame.getBoardGameName(), rateBoardGame.getRate());

        log.debug("Saving new rating entity for boardgame {}", rateBoardGame.getBoardGameName());

        rateRepository.save(build);

        log.debug("Return updated rate for boardgame {}", rateBoardGame.getBoardGameName());

        return boardGameRepository.findByName(rateBoardGame.getBoardGameName())
                .map(this::ratedBoardGame)
                .orElseThrow(() -> new BoardGameNotFound(rateBoardGame.getBoardGameName()));
    }

    List<BoardGame> withHigherRateThan(Double rate) {
        log.debug("Find all boardgames with rate higher than {}", rate);

        return boardGameRepository.findAll()
                .stream()
                .map(this::ratedBoardGame)
                .filter(game -> game.getRating().getAverageRate() > rate)
                .collect(Collectors.toList());
    }

    private BoardGame ratedBoardGame(BoardGameEntity bg) {
        log.debug("Retrieving all the rates of boardgame {}", bg.getName());

        final List<RateEntity> ratesOfTheBoardGame = rateRepository.findByBoardGameName(bg.getName());
        final Double average = ratesOfTheBoardGame.stream()
                .mapToDouble(RateEntity::getRate)
                .average()
                .getAsDouble();

        return new BoardGame(bg.getName(), Category.valueOf(bg.getCategory().name()),
                new Rating(average),
                new AgeRange(bg.getMinimalAge(), bg.getMaximalAge()),
                new NumberOfPlayers(bg.getMinimalNumberOfPlayers(), bg.getMaximalNumberOfPlayers())
        );
    }
}

File: src/main/java/com/openvalue/boardgameratings/service/rating/BoardGameRatingService.kt
package com.openvalue.boardgameratings.service.rating

import com.openvalue.boardgameratings.api.*
import com.openvalue.boardgameratings.service.boardgame.BoardGameEntity
import com.openvalue.boardgameratings.service.boardgame.BoardGameNotFound
import com.openvalue.boardgameratings.service.boardgame.BoardGameRepository
import mu.KotlinLogging
import org.springframework.stereotype.Service

@Service
open class BoardGameRatingService (
        private val boardGameRepository: BoardGameRepository,
        private val rateRepository: RateRepository
) {

    private val logger = KotlinLogging.logger {  }

    @Throws(BoardGameNotFound::class)
    fun ratingBoardGame(rateBoardGame: RateBoardGame): BoardGame {
        val build = RateEntity(null, rateBoardGame.boardGameName, rateBoardGame.rate)

        logger.debug("Saving new rating entity for boardgame ${rateBoardGame.boardGameName}")

        rateRepository.save(build)

        logger.debug("Return updated rate for boardgame ${rateBoardGame.boardGameName}")

        return boardGameRepository.findByName(rateBoardGame.boardGameName)
                .map { this.ratedBoardGame(it) }
                .orElseThrow { BoardGameNotFound(rateBoardGame.boardGameName) }
    }

    fun withHigherRateThan(rate: Double): List<BoardGame> {
        logger.debug("Find all boardgames with rate higher than $rate")

        return boardGameRepository.findAll()
                .map{ this.ratedBoardGame(it) }
                .filter { it.rating.averageRate > rate }
    }
    private fun ratedBoardGame(bg: BoardGameEntity): BoardGame {
        logger.debug("Retrieving all the rates of boardgame ${bg.name}")

        val ratesOfTheBoardGame = rateRepository.findByBoardGameName(bg.name)
        val average = ratesOfTheBoardGame.asSequence().map { it.rate }.average()

        return BoardGame(bg.name, Category.valueOf(bg.category.name),
                AgeRange(bg.minimalAge, bg.maximalAge),
                NumberOfPlayers(bg.minimalNumberOfPlayers, bg.maximalNumberOfPlayers),
                Rating(average)
        )
    }

}
